idx/memory/partitioned_idx.go

```diff
@@ -0,0 +1,536 @@
+func (p *PartitionedMemoryIdx) GetPath(orgId uint32, path string) []idx.Archive {
+	g, _ := errgroup.WithContext(context.Background())
+	result := make([][]idx.Archive, len(p.Partition))
+	var i int
+	for _, m := range p.Partition {
+		pos, m := i, m
+		g.Go(func() error {
+			result[pos] = m.GetPath(orgId, path)
+			return nil
+		})
+		i++
+	}
+	if err := g.Wait(); err != nil {
+		log.Errorf("memory-idx: failed to getPath: orgId=%d path=%s. %s", orgId, path, err)
+		return nil
+	}
+
+	// get our total count, so we can allocate our response in one go.
+	items := 0
+	for _, r := range result {
+		items += len(r)
+	}
+	response := make([]idx.Archive, 0, items)
+	for _, r := range result {
+		response = append(response, r...)
+	}
+	return response
+}
+
+// Delete deletes items from the index
+// If the pattern matches a branch node, then
+// all leaf nodes on that branch are deleted. So if the pattern is
+// "*", all items in the index are deleted.
+// It returns a copy of all of the Archives deleted.
+func (p *PartitionedMemoryIdx) Delete(orgId uint32, pattern string) ([]idx.Archive, error) {
+	g, _ := errgroup.WithContext(context.Background())
+	result := make([][]idx.Archive, len(p.Partition))
+	var i int
+	for _, m := range p.Partition {
+		pos, m := i, m
+		g.Go(func() error {
+			deleted, err := m.Delete(orgId, pattern)
+			if err != nil {
+				return err
+			}
+			result[pos] = deleted
+			return nil
+		})
+		i++
+	}
+	if err := g.Wait(); err != nil {
+		log.Errorf("memory-idx: failed to Delete: orgId=%d pattern=%s. %s", orgId, pattern, err)
+		return nil, err
+	}
+
+	// get our total count, so we can allocate our response in one go.
+	items := 0
+	for _, r := range result {
+		items += len(r)
+	}
+	response := make([]idx.Archive, 0, items)
+	for _, r := range result {
+		response = append(response, r...)
+	}
+	return response, nil
+}
+
+// Find searches the index for matching nodes.
+// * orgId describes the org to search in (public data in orgIdPublic is automatically included)
+// * pattern is handled like graphite does. see https://graphite.readthedocs.io/en/latest/render_api.html#paths-and-wildcards
+// * from is a unix timestamp. series not updated since then are excluded.
+func (p *PartitionedMemoryIdx) Find(orgId uint32, pattern string, from int64) ([]idx.Node, error) {
+	g, _ := errgroup.WithContext(context.Background())
+	resultChan := make(chan []idx.Node)
+	for _, m := range p.Partition {
+		m := m
+		g.Go(func() error {
+			//fmt.Printf("running find on %d\n", pos)
+			found, err := m.Find(orgId, pattern, from)
+			if err != nil {
+				return err
+			}
+			resultChan <- found
+			//fmt.Printf("completed find on %d\n", pos)
+			return nil
+		})
+	}
+
+	// consume from the ResultChan and merge the results, de-duping branches
+	byPath := make(map[string]*idx.Node)
+	done := make(chan struct{})
+	go func() {
+		for found := range resultChan {
+			for _, node := range found {
+				if _, ok := byPath[node.Path]; !ok {
```

Comment:
if instead of `_` this would assign the value to a variable then this variable could be used in the `if` / `else` block instead of doing `byPath[node.Path]` many times. if the result set is large we can save a bunch of lookups in large `byPath` maps by doing that.